### PR TITLE
Fix/removes use of deprecated req.param method

### PR DESCRIPTION
**Pull request by aars**
Removes use of deprecated req.param() method.

req.param() would look in req.params, req.query and req.body. I tried to make logical choices on where to look for certain params (e.g. don't lookup redirect parameter in request body). Personally I would not allow an access-token in the request body and only as a query parameter or header, but people might depend on it so I kept it.

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by aars** [comment]
I'm not sure how to handle the appVeyor error.

File: common/models/access-token.js [hunk 1]
```diff
@@ -171,7 +171,8 @@ module.exports = function(AccessToken) {
     cookies = cookies.concat(['access_token', 'authorization']);
 
     for (length = params.length; i < length; i++) {
-      id = req.param(params[i]);
+      // Only look up params in query. There is no body here.
+      id = req.query && req.query[params[i]];
```
Comment (bajtos) [inline common/models/access-token.js]: IMO you should check `req.params` too.

File: lib/application.js [hunk 2]
```diff
@@ -296,7 +296,9 @@ app.enableAuth = function() {
     var req = ctx.req;
     var Model = method.ctor;
     var modelInstance = ctx.instance;
-    var modelId = modelInstance && modelInstance.id || req.param('id');
+    var modelId = (modelInstance && modelInstance.id) ||
+      (req.params && req.params.id) ||
+      (req.body && req.body.id);
```
Comment (bajtos) [inline lib/application.js]: Note that `req.param` checked `req.body` before `req.params` as can be seen [here](https://github.com/strongloop/express/blob/4db51fbeba0cd543949e243672c0e6c3c3c70639/lib/request.js#L213-L215).

Please add a unit-test to prevent such regressions from resurfacing in the future.

Comment (bajtos) [inline lib/application.js]: Correction: `req.body` before `req.query`

File: common/models/access-token.js (hunk 1 above)
Comment (dougwilson) [inline common/models/access-token.js]: And, really, unless this is used outside of Express, `req.query` is guaranteed to be an object.

File: lib/application.js (hunk 2 above)
Comment (dougwilson) [inline lib/application.js]: Hopefully `req.params` should be guaranteed to be an object at this location.

Comment (aars) [inline lib/application.js]: I put it at the bottom because `req.param()` does indeed check the body first, but the id should not be in the request body. Should both `req.params.id` and `req.body.id` I believe the body should be ignored.

Why not remove `req.body.id`?

File: common/models/access-token.js (hunk 1 above)
Comment (aars) [inline common/models/access-token.js]: Ok, I'll remove the check on `req.query`.

File: lib/application.js (hunk 2 above)
Comment (aars) [inline lib/application.js]: And since `req.params` defaults to `{}` I'd suggest simply:
``` js
var modelId = (modelInstance) ? modelInstance.id : req.params.id;
```

File: common/models/access-token.js (hunk 1 above)
Comment (aars) [inline common/models/access-token.js]: I'm refactoring this, but there is one test-case that mocks a request that does _not_ contain `req.params` or `req.query`. I've contacted @bajtos since it's his test.

File: lib/application.js (hunk 2 above)
Comment (bajtos) [inline lib/application.js]: > but the id should not be in the request body

That's AFAIK not true, e.g. `updateOrCreate` takes the id from the request body.

More importantly, we must preserve backwards compatibility with previous loopback versions. So even if the `id` was not supposed to be read from the body, I would argue that we should still check the body for the sake of staying backwards compatible.

Comment (aars) [inline lib/application.js]: `updateOrCreate` is a good point! Though I wonder if a requestor should ever be able to determine a `modelId`, but that is another discussion. (Backwards compatibility is not always a good thing) Let's keep the `req.body` lookup.

@bajtos Do you agree that `req.params.id` should take precedence over `req.body.id`?

Comment (bajtos) [inline lib/application.js]: >  Do you agree that `req.params.id` should take precedence over `req.body.id`?

Yes, we must preserve the behaviour of `req.param()`.

File: common/models/user.js [hunk 3]
```diff
@@ -564,7 +564,11 @@ module.exports = function(User) {
     UserModel.on('attached', function() {
       UserModel.afterRemote('confirm', function(ctx, inst, next) {
         if (ctx.req) {
-          ctx.res.redirect(ctx.req.param('redirect'));
+          // Find redirect parameter as param or query. Not in body.
+          var redirectTo = (ctx.req.params && ctx.req.params.redirect) ||
+            (ctx.req.query && ctx.req.query.redirect);
```
Comment (bajtos) [inline common/models/user.js]: Why the redirect parameter cannot be in the body? Unless you can prove that such request would not be valid, or that such usage is explicitly forbidden by the docs, then this patch must check `req.body` too in the same way as `req.param()` does.

File: lib/application.js [hunk 4]
```diff
@@ -295,13 +295,20 @@ app.enableAuth = function() {
...
     var modelSettings = Model.settings || {};
-    var errStatusCode = modelSettings.aclErrorStatus || app.get('aclErrorStatus') || 401;
-    if (!req.accessToken) {
-      errStatusCode = 401;
+
+    var errStatusCode = 401;
+    if (req.accessToken) {
+      errStatusCode = modelSettings.aclErrorStatus ||
+                      app.get('aclErrorStatus')    ||
+                      errStatusCode;
```
Comment (bajtos) [inline lib/application.js]: What is the purpose of this change? It looks unrelated to `req.param()` fixes.

**Comment by bajtos** [comment]
> I'm not sure how to handle the appVeyor error.

Please ignore that, @cgole is working on disabling that integration.

**Comment by bajtos** [comment]
Closing in favour of #1038 to get this fixed ASAP.
